### PR TITLE
refactor everything!!

### PR DESCRIPTION
**Pull request by antonbobkov**
put pending_transactions into blocktree

### CODE REVIEW

**Comment by antonbobkov** [comment]
oops, forgot to do a different branch

just look at the last commit, I added orphan blocks

still need to do tests and orphan blocks auto-add

**Comment by antonbobkov** [comment]
last commit is actually interesting :P

File: src/blocktree.erl [hunk 1]
```diff
@@ -24,80 +31,34 @@
...
-	    pass;
-	error ->
-	    TransactionArray = array:new()
-    end,
-
-    {Msg, NewTA} = add_new_transaction_to_array(Transaction, TransactionArray),
-    NewTM = maps:put(Id, NewTA, TM),
-    {Msg, NewTM}.
+%% @doc Initializes an empty container
+-spec init() -> treedata().
```
Comment (pdlla) [inline src/blocktree.erl]: :o 👍

File: src/blocktree.erl [hunk 2]
```diff
@@ -24,80 +31,34 @@
...
+init() ->
```
Comment (pdlla) [inline src/blocktree.erl]: how about calling this new_treedata() instead?

File: src/pop_protocol.erl [hunk 3]
```diff
@@ -363,3 +376,64 @@ generate_new_block(VerifierIndex, PD)
...
+	{ok, _} ->
+	    {ignored_duplicate, ProtocolData};
+	error ->
+	    R2 = find_block_by_id(maps:get(PrevId, Block), ProtocolData),
+	    case R2 of
+		{ok, _} ->
+		    {added_new, add_block_in_order(Block, CurrentTime, ProtocolData)};
+		error ->
+		    OrphanMapNew = maps:put(ThisId, Block, OrphanMap),
+		    {orphan, ProtocolData#protocol_data{orphan_blocks = OrphanMapNew}}
```
Comment (pdlla) [inline src/pop_protocol.erl]: nice!

File: src/blocktree.erl (hunk 2 above)
Comment (antonbobkov) [reply src/blocktree.erl]: YA

Comment (antonbobkov) [reply src/blocktree.erl]: wait no

Comment (antonbobkov) [reply src/blocktree.erl]: maybe?

Comment (antonbobkov) [reply src/blocktree.erl]: how about just new()?

blocktree:new()

File: src/pop_protocol.erl (hunk 3 above)
Comment (pdlla) [reply src/pop_protocol.erl]: I think "orphan" is actually inconsistent with the usage of the term in bitcoin.

Here orphan means there is no parents

in bitcoin, all orphan blocks have parents but are not part of the longest chain

😱

File: src/blocktree.erl (hunk 2 above)
Comment (pdlla) [reply src/blocktree.erl]: anything but init 😱

File: src/pop_protocol.erl (hunk 3 above)
Comment (antonbobkov) [reply src/pop_protocol.erl]: `all orphan blocks have parents`

ಠ_ಠ

Comment (pdlla) [reply src/pop_protocol.erl]: :o?

Comment (pdlla) [reply src/pop_protocol.erl]: I guess it's like they all got abandoned by their parent in the longest chain 😱😱😱😱

**Pull request merged**